src/coreclr/jit/promotion.cpp

```diff
@@ -0,0 +1,1310 @@
+
+                if (lcl->OperIs(GT_LCL_ADDR))
+                {
+                    assert(user->OperIs(GT_CALL) && dsc->IsHiddenBufferStructArg() &&
+                           (user->AsCall()->gtArgs.GetRetBufferArg()->GetNode() == lcl));
+
+                    accessType   = TYP_STRUCT;
+                    accessLayout = m_compiler->typGetObjLayout(user->AsCall()->gtRetClsHnd);
+                    accessFlags  = AccessKindFlags::IsCallRetBuf;
+                }
+                else
+                {
+                    accessType   = lcl->TypeGet();
+                    accessLayout = accessType == TYP_STRUCT ? lcl->GetLayout(m_compiler) : nullptr;
+                    accessFlags  = ClassifyLocalRead(lcl, user);
+                }
+
+                LocalUses* uses = GetOrCreateUses(lcl->GetLclNum());
+                unsigned   offs = lcl->GetLclOffs();
+                uses->RecordAccess(offs, accessType, accessLayout, accessFlags, m_curBB->getBBWeight(m_compiler));
+            }
+        }
+
+        return fgWalkResult::WALK_CONTINUE;
+    }
+
+private:
+    //------------------------------------------------------------------------
+    // GetOrCreateUses:
+    //   Get the uses information for a local. Create it if it does not already exist.
+    //
+    // Parameters:
+    //   lclNum - The local
+    //
+    // Returns:
+    //   Uses information.
+    //
+    LocalUses* GetOrCreateUses(unsigned lclNum)
+    {
+        if (m_uses[lclNum] == nullptr)
+        {
+            m_uses[lclNum] = new (m_compiler, CMK_Promotion) LocalUses(m_compiler);
+        }
+
+        return m_uses[lclNum];
+    }
+    //------------------------------------------------------------------------
+    // ClassifyLocalAccess:
+    //   Given a local use and its user, classify information about it.
+    //
+    // Parameters:
+    //   lcl - The local
+    //   user - The user of the local.
+    //
+    // Returns:
+    //   Flags classifying the access.
+    //
+    AccessKindFlags ClassifyLocalRead(GenTreeLclVarCommon* lcl, GenTree* user)
+    {
+        assert(lcl->OperIsLocalRead());
+
+        AccessKindFlags flags = AccessKindFlags::None;
+        if (user->IsCall())
+        {
+            GenTreeCall* call     = user->AsCall();
+            unsigned     argIndex = 0;
+            for (CallArg& arg : call->gtArgs.Args())
+            {
+                if (arg.GetNode() != lcl)
+                {
+                    argIndex++;
+                    continue;
+                }
+
+                flags |= AccessKindFlags::IsCallArg;
+
+                unsigned argSize = 0;
+                if (arg.GetSignatureType() != TYP_STRUCT)
+                {
+                    argSize = genTypeSize(arg.GetSignatureType());
+                }
+                else
+                {
+                    argSize = m_compiler->typGetObjLayout(arg.GetSignatureClassHandle())->GetSize();
+                }
+
+#ifdef WINDOWS_AMD64_ABI
```

Comment:
Leverage `getArgTypeForStruct` here?

Reply:
I will remove these for now, they are currently unused. The plan was to move ABI information determination back so that it happened before this pass and could be used for it.